File: frontend/src/api/pipeline.ts
import client from './client';

export interface PipelineRunPayload {
    goal: string;
    cv_text: string;
    skills: string[];
}

export interface PipelineRunSummary {
    id: string;
    label: string;
    created_at: string | null;
    ats_score: number | null;
    status: string;
    current_stage: number;
}

export interface PipelineRunsResponse {
    runs: PipelineRunSummary[];
}

/** Full pipeline state as returned by GET /api/pipeline/:id/result (state_json). */
export interface PipelineResultState {
    pipeline_id?: string;
    user_id?: string;
    status?: string;
    current_stage?: number;
    cv_raw?: string;
    job_description?: string;
    candidate_profile?: Record<string, unknown>;
    ats_score?: number | null;
    ats_breakdown?: Record<string, unknown> | null;
    skill_gaps?: string[] | null;
    skill_match_score?: number | null;
    market_trends?: Record<string, unknown> | null;
    salary_benchmarks?: Record<string, unknown> | null;
    cv_critique?: string | null;
    optimised_cv?: string | null;
    cover_letter?: string | null;
    job_tier?: string | null;
    missing_skills?: string[] | null;
    skill_roadmap?: Array<{ skill?: string; weeks?: number; completed?: boolean; focus?: string; duration_weeks?: number; milestones?: string[] }> | null;
    interview_question_bank?: string[] | null;
    market_analysis?: { market_analysis?: Record<string, { status?: string; snippets?: string[] }>; hot_skills?: string[] };
    created_at?: string;
    completed_at?: string | null;
    error_log?: string[];
    critique?: {
        summary?: string;
        matching_skills?: string[];
        transferable_skills?: string[];
        missing_critical_skills?: string[];
    };
    implicit_skills?: string[];
}

export async function runPipeline(payload: PipelineRunPayload) {
    const { data } = await client.post('/pipeline/start', {
        cv_text: payload.cv_text,
        job_description: payload.goal,
        options: {
            run_interview_prep: true,
            tone: "formal"
        }
    });
    return data as { status: string; pipeline_id: string };
}

export async function getPipelineRuns(limit = 8): Promise<PipelineRunsResponse> {
    const { data } = await client.get<PipelineRunsResponse>('/pipeline/runs', { params: { limit } });
    return data;
}

export async function getPipelineStatus(pipelineId: string) {
    const { data } = await client.get(`/pipeline/${pipelineId}/status`);
    return data as {
        status: string;
        current_stage: number;
        completed_stages: number[];
        error_log: string[];
    };
}

export async function getPipelineResult(pipelineId: string): Promise<PipelineResultState> {
    const { data } = await client.get<PipelineResultState>(`/pipeline/${pipelineId}/result`);
    return data;
}

/** @deprecated Use getPipelineStatus */
export async function getTaskState(pipelineId: string) {
    return getPipelineStatus(pipelineId);
}

File: backend/app/schemas/pipeline.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel
import uuid
from datetime import datetime

class PipelineStartOptions(BaseModel):
    run_interview_prep: bool = True
    tone: str = "formal"


class PipelineStartRequest(BaseModel):
    cv_text: str
    job_description: str
    options: Optional[PipelineStartOptions] = None


class PipelineInputRequest(BaseModel):
    job_description: Optional[str] = None
    cv_raw: Optional[str] = None
    skills: Optional[List[str]] = None


# ─── Response Models ────────────────────────────────────────────────────────

class PipelineRunSummary(BaseModel):
    id: uuid.UUID
    label: str
    created_at: Optional[datetime] = None
    ats_score: Optional[float] = None
    status: str
    current_stage: int


class PipelineListResponse(BaseModel):
    runs: List[PipelineRunSummary]
    total: int
    skip: int
    limit: int


class PipelineStatusResponse(BaseModel):
    status: str
    current_stage: int
    completed_stages: List[int]
    error_log: List[str]


class PipelineResultState(BaseModel):
    model_config = {"extra": "ignore"}
    
    pipeline_id: Optional[uuid.UUID] = None
    user_id: Optional[uuid.UUID] = None
    status: Optional[str] = None
    current_stage: Optional[int] = None
    cv_raw: Optional[str] = None
    job_description: Optional[str] = None
    preferred_tone: Optional[str] = None
    candidate_profile: Optional[Dict[str, Any]] = None
    ats_score: Optional[float] = None
    ats_breakdown: Optional[Dict[str, Any]] = None
    skill_gaps: Optional[List[str]] = None
    skill_match_score: Optional[float] = None
    implicit_skills: Optional[List[str]] = None
    market_analysis: Optional[Dict[str, Any]] = None
    salary_benchmarks: Optional[Dict[str, Any]] = None
    critique: Optional[Dict[str, Any]] = None
    cv_critique: Optional[str] = None  # Legacy support
    optimised_cv: Optional[Any] = None # Can be string or Dict
    cover_letter: Optional[str] = None
    job_tier: Optional[str] = None
    missing_skills: Optional[List[str]] = None
    skill_roadmap: Optional[List[Any]] = None
    interview_question_bank: Optional[List[Any]] = None
    messages: Optional[List[str]] = None
    missing_fields: Optional[List[str]] = None
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_log: Optional[List[str]] = None
